### ISSUE TITLE
Named Views in ElasticSearch as an alternative to Filtered Aliases

### ISSUE THREAD

**Issue by tmpaul06** [elastic/elasticsearch#25588]
## Named views

I am creating ticket to discuss the possibility of creating "named" views in ElasticSearch. I'm aware of filtered aliases, but ran into a lot of problems with the approach. I will try to express the need for such a feature with an use case that I have encountered in the past.

### Use case

Given an external input of a large number of id or id like fields, define a view over an ElasticSearch index. The external input can be a CSV file, or a JSON passed in from another database. It can be a terms filter or an ids filter or a more complex combination of other filters.

I am highlighting some discussions from ES discuss threads other than my own to support the above mentioned use case.
* https://github.com/elastic/elasticsearch/issues/23281
* https://discuss.elastic.co/t/filtered-aliases-and-huge-ids-terms-filter/58844

### Why not use aliases ?

There are two main reasons why I think aliases should not be used as views.

1. The first one is conceptual. An alias is typically used to point to a different version of the "same" index. It could be new changes to the data, adding more documents, but a filtered subset of the index is conceptually not an `alias` of the index.

    Also if aliases are used as views, hitting the index alias query will return the filtered aliases for that index. If there are several filtered aliases created for a single index, each with millions of fields, the get_alias request will explode in terms of size. (Ran into this problem with ES 5.2.2)

    We faced this exact issue with large terms filter using filtered aliases.

2. Different views can be computed on a single index alias. This allows having two different computed views of historical and current data side-by-side. A view is more granular than an alias. A filter created for a view can be combined with several other filters to perform set like operations on the index.

### Assumptions
1. The filter can be cached. But there is no guarantee that it will remain in memory during query execution. If evicted, it will simply be recomputed similar to existing filters.

2. The view will be computed infrequently, i.e the filters themselves will change infrequently. In the event of a change in the filter, the view will be recomputed (by making an explicit call to ES hereafter called the **update_view** operation or during query time).

3. The filter can contain a large number of term or id based field values. If it is too big to fit in one request, the filter can be computed with batched requests. (Similar to update_view).

4. Each cached filter will be for a given index and doc_type.

5. A view can be queried against using a named filter called **view query**
```javascript
// GET /_search
{
    "query": {
        "constant_score" : {
            "filter" : {
                "view" : "my_named_view"
            }
        }
    }
}
```

or by explicitly quering against a view
```
curl -XGET /_view/my_named_view -d ....

and

curl -XPUT /_view/my_named_view - d... // Updated filter
```

### Questions
1. It is possible to use the existing filter cache to implement this ?  A named view would simply be then a named filter. The name references the stored/cached bitset.
2. Are there any problems/challenges due to the distributed nature ?

In short, I would like to create a subset of an index and refer to it by name in queries rather than sending in large term filters for each query. There is no additional metadata that I can add to the documents to avoid sending the large number of term filters, because the collection of the values in the filter are created externally, e.g from a CSV file, or even a random sample of id fields in the index.

I understand that creating filtered aliases with large number of term filters is kind of an anti-pattern, but the alternative is to create a separate index for a given subset of documents. The reindexing time + space used by the document subset is prohibitive for our use case.

Current ES version: `5.2.0`

Does this look like an sensible idea ?

**Comment by jimczi** [comment]
> It is possible to use the existing filter cache to implement this ? A named view would simply be then a named filter. The name references the stored/cached bitset.

No, the filter cache is per segment so you cannot expect to cache your view in a single call. It should be recomputed on each newly added segment (and after merges, ...).
Creating a `view` of a particular filter in a single call is only valid if your index does not change.
The filter cache does not contain the document itself but just a pointer to the internal id of the documents that match the filter in the segment. The entry is valid for this segment only and when the documents are updated (or added) a new entry needs to be computed in the new segment.

So If your main problem is just a slow `terms` query then you should try another approach and follow the advices given in the discussions that you linked.
Giant queries with lots of terms that try to return documents based on their _id is clearly an anti pattern in ES and should be avoided. They are plenty of solutions to circumvent this problem so I'll close this issue for now and we can discuss in the forum if you want. Just open a new thread describing your problem and I'll be happy to help.
